File: 단계별로 풀어보기/Step 15/14888.cpp
#include <iostream>
#include <vector>

using namespace std;

int N;
int minNum=1000000000, maxNum=-1000000000;
vector<int> num;
vector<int> oper;

void DFS(int result, int idx, int sum, int sub, int mul, int div) {
    if(idx == N) {
        maxNum = max(maxNum, result);
        minNum = min(minNum, result);
        return;
    }

    if(sum!=0){
        DFS(result+num[idx], idx+1, sum-1, sub, mul, div);
    }
    if(sub!=0){
        DFS(result-num[idx], idx+1, sum, sub-1, mul, div);
    }
    if(mul!=0){
        DFS(result*num[idx], idx+1, sum, sub, mul-1, div);
    }
    if(div!=0){
        DFS(result/num[idx], idx+1, sum, sub, mul, div-1);
    }
}

int main() {
    int i, tmp;

    cin >> N;
    for(i=0; i<N; i++) {
        cin >> tmp;
        num.push_back(tmp);
    }
    for(i=0; i<4; i++) {
        cin >> tmp;
        oper.push_back(tmp);
    }

    DFS(num[0], 1, oper[0], oper[1], oper[2], oper[3]);

    cout << maxNum << endl << minNum;
    return 0;
}

File: StepByStep/Python/Step_15_백트래킹/14888.py
import sys

N = int(input())
nums = list(map(int, sys.stdin.readline().split()))
ops = list(map(int, sys.stdin.readline().split()))
result_list = []


def calculator(result, idx, add, sub, mul, div):
    if idx == N:
        result_list.append(result)
        return

    if add != 0:
        calculator(result+nums[idx], idx+1, add-1, sub, mul, div)
    if sub != 0:
        calculator(result-nums[idx], idx+1, add, sub-1, mul, div)
    if mul != 0:
        calculator(result*nums[idx], idx+1, add, sub, mul-1, div)
    if div != 0:
        calculator(-(-result // nums[idx]) if result < 0 else result // nums[idx], idx+1, add, sub, mul, div-1)


calculator(nums[0], 1, ops[0], ops[1], ops[2], ops[3])
print(max(result_list))
print(min(result_list))
